ES|QL rule docs (#4062)
elastic/security-docs: docs/detections/rules-ui-create.asciidoc

{esql} rule queries are loosely categorized into two types: aggregating and non-aggregating. 

[float]
[[esql-agg-query]]
===== Aggregating query 

Aggregating queries use {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate source event data. Alerts generated by an {esql} rule with an aggregating query only contain the fields returned by the query.

Here is an example aggregating query:

[source,esql]
----
FROM logs-*
| STATS host_count = COUNT(host.name) BY host.name
| SORT host_count DESC
| WHERE host_count > 20
----

- This query starts by searching logs from indices that match the pattern `logs-*`. 
- The query then aggregates the count of events by `host.name`.
- Next, it sorts the result by `host_count` in descending order.
- Then, it filters for events where the `host_count` field appears more than 20 times during the specified rule interval.

NOTE: Rules that use aggregating queries might create duplicate alerts. This can happen  when events that occur in the additional look-back time are aggregated both in the current rule execution and in a previous rule execution.

[float]
[[esql-non-agg-query]]
===== Non-aggregating query 

> A non-aggregating query doesn't use `STATS...BY` functions and doesn't aggregate source event data. Generated alerts contain source event data and fields returned by the {esql} query. 

Author: ```suggestion
A non-aggregating query doesn't use `STATS...BY` functions and doesn't aggregate source event data. Generated alerts contain source event data and fields returned by the {esql} query. 

Non-aggregating queries doesn't use `STATS...BY` functions and doesn't aggregate source event data. Alerts generated by an {esql} rule with a non-aggregating query only contain the fields returned by the query.
```

>  Non-aggregating queries doesn't use `STATS...BY` functions and doesn't aggregate source event data. Alerts generated by an {esql} rule with a non-aggregating query only contain the fields returned by the query.

Reviewer: ```suggestion
Non-aggregating queries doesn't use `STATS...BY` functions and doesn't aggregate source event data. Alerts generated by an {esql} rule with a non-aggregating query only contain the fields returned by the query.
```
This doesn't render correctly, which I think may be caused by this space.

Author: Yeah, I was just about to fix this copy/paste error. Thank you!

[discrete]
[[create-esql-rule]]
=== Create an {esql} rule

preview::[]

Use {ref}/esql.html[{esql}] to query your source events and aggregate event data. Query results are returned in a table with rows and columns. Each row becomes an alert.

To create an {esql} rule:

. Go to *Rules* -> *Detection rules (SIEM)* -> *Create new rule*. The *Create new rule* page appears.
. Select **{esql}**, then write a <<esql-rule-query-types,query>>. 
+
TIP: Click the help icon (image:images/esql-ref-button.png[Click the ES|QL help icon,20,20]) to open the in-product reference documentation for all {esql} commands and functions.
+

> NOTE: To avoid conflicts between the {ref}/esql-commands.html#esql-limit[`LIMIT` value] (which defines how many rows are returned in the results table) and the <<opt-fields-all,`max_signals`>> value (which determines the max number of alerts that can be generated per rule execution), make sure the `max_signals` value does not exceed the `LIMIT` value. If you don't define a `LIMIT` value, the rule automatically uses the `max_signals` value. 

Author: @vitaliidm let me know if this note needs any tweaks. Thanks in advance for the feedback!

[discrete]
[[create-esql-rule]]
=== Create an {esql} rule

preview::[]

Use {ref}/esql.html[{esql}] to query your source events and aggregate event data. Query results are returned in a table with rows and columns. Each row becomes an alert.

To create an {esql} rule:

. Go to *Rules* -> *Detection rules (SIEM)* -> *Create new rule*. The *Create new rule* page appears.
. Select **{esql}**, then write a <<esql-rule-query-types,query>>. 
+
TIP: Click the help icon (image:images/esql-ref-button.png[Click the ES|QL help icon,20,20]) to open the in-product reference documentation for all {esql} commands and functions.
+

> [NOTE] 

Author: @vitaliidm let me know if this note needs any tweaks. Thanks in advance for the feedback!

Author: Thanks for the prompt feedback, @vitaliidm!

RE: Your first point about `max_signals` not being configurable through the UI:  

The ability to only configure `max_signals` through the API, and documenting this information there was sorta what I was alluding to [here](https://github.com/elastic/security-docs/pull/4062#discussion_r1374571790). It's also why I linked to the `max_signals` docs in the API docs, but it might be too presumptuous to assume that users would make the connection that the `max_signals` field is only configurable via the Create rule API. 

The clearest and most helpful way to doc this information is to put it where users need it, which is in the API docs. I still think the `max_signals` [docs](https://www.elastic.co/guide/en/security/master/rules-api-create.html#opt-fields-all) is the appropriate place to put this information, even though the description for that field applies generally to all rule types. Adding this information as a note shows that there's a caveat to the functionality that users should be aware of when they're working with a specific rule.

Author: @vitaliidm I'll merge this PR and let's continue this convo over in https://github.com/elastic/security-docs/issues/4195.

[[create-esql-rule]]
=== Create an ES|QL rule

preview::[]

. Go to *Rules* -> *Detection rules (SIEM)* -> *Create new rule*. The *Create new rule* page displays.
. Select **ES|QL**. 
. Write an aggregating or non-aggregating type query:
+
TIP: Click the **ES|QL reference** button [book icon] for more information on ES|QL syntax. 
+

> ** **Aggregating query**: Uses {ref}/esql-agg-functions.html[STATS...BY] functions to aggregate data. Results from each query row are not saved to an alert document within Elasticsearch. 

Author: ```suggestion
** **Aggregating query**: Uses {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate data. Results from each query row are not saved to an alert document within Elasticsearch. 
```

[[create-esql-rule]]
=== Create an ES|QL rule

preview::[]

. Go to *Rules* -> *Detection rules (SIEM)* -> *Create new rule*. The *Create new rule* page displays.
. Select **ES|QL**. 

> . Write an aggregating or non-aggregating type query:

Author: Yeah, I was thinking about making this more general too. Just in case we want to specify more query types in the future. How about something like this:

```suggestion
. Write an ES|QL query. You can create any of the following query types:
```

> . Write an aggregating or non-aggregating type query:

Author: Resolving because I responded [here](https://github.com/elastic/security-docs/pull/4062#discussion_r1371772895).

> . Write an aggregating or non-aggregating type query:

Reviewer 2: I don't think we need event this: `You can create any of the following query types:`
Just ` Write an ES|QL query`

+

> TIP: Click the **ES|QL reference** button [book icon] for more information on ES|QL syntax. 

Reviewer 2: should we specify button located in ES|QL input or add a screenshot with it?

Author: Yep! Added these deets in [ab28d69](https://github.com/elastic/security-docs/pull/4062/commits/ab28d69cc19c54046fda4c1e54df7dee78bf0ab6).

+

> ** **Aggregating query**: Uses {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate alert data. An alert is generated for each segment of the query, but is not associated with a specific source event since aggregating queries process data from multiple alert documents. As a result, alerts from aggregating queries do not contain source event data.

Reviewer 2: > An alert is generated for each segment of the query,

What is meant by `each segment of the query`?

Reviewer 2: I would try to rephrase it as: Generated alerts contain only fields returned by query, without mentioning source event documents. Let's leave it for non-aggregating query only,

Author: @vitaliidm to clarify, are you suggesting that the second and third sentences be revised like so:

```suggestion
** **Aggregating query**: Uses {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate alert data. An alert is generated for each row in the result table, and alert documents only contain fields returned by the query.  
```

+
Example aggregating query: 
+
[source,esql]
----
FROM logs*
| STATS count = COUNT(host.name) BY host.name

> | SORT host.name

Author: @vitaliidm I was thinking we could frame these examples in a more descriptive way. Maybe something like:

For example, if you wanted to create an alert for < describe scenario >, you can write a query like this:

> | SORT host.name

Reviewer 2: We can describe this query as: 

> Rule with the following query generates alert for each host.name that encountered more than 20 times during rule interval.

For more production oriented rules, I think we would need to contact TRADE team. But, it can be very complex, bussiness specific rule. I a m not sure we want to add it in our docs, that should have simple enough examples to give basic understanding how rules work

> | SORT host.name

Author: Thanks for the description! That was the main item I needed for this part :)

----

> ** **Non-aggregating query**: Doesn't use `STATS...BY` functions and doesn't aggregate alert data. An alert is generated for each segment of the query, and each alert is associated with a source event. As a result, alerts from non-aggregating queries contain source event data.

Reviewer 2: > functions and doesn't aggregate alert data

```suggestion
** **Non-aggregating query**: Doesn't use `STATS...BY` functions and doesn't aggregate source data. An alert is generated for each row in the result, and each alert can be linked to a source event through [metadata] operator. As a result, alerts from non-aggregating queries contain source event data merged with fields returned by query request
```

Reviewer 2: row, column, table - something described in our tech docs: https://esql.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/esql.html
So, we can use it our advantage and add link to it in our doc.
And we can put in common area of ES|QL section: each row become an alert. That is cornerstone for how this rule type works

Author: Would you mind elaborating on the last part of your comment? I'm not clear on whether you're suggesting that we explain how the ES|QL rule generates alerts in these docs, or in the [core ES|QL docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/esql.html).

Author: @vitaliidm in addition to explaining how ES|QL alerts are generated, do you think we need to mention any best practices for writing ES|QL rule queries? For example, if users drop certain fields in the query, will those fields be omitted from the alert document? Also, should they include certain fields to ensure they can use specific functionality in the alert details flyout, like the [Entities](https://www.elastic.co/guide/en/security/master/view-alert-details.html#entities-overview) overview?

Author: Oh yeah, I'm not suggesting that we go through all of our features to find fields that users need to keep or return. I'm more so asking if we need to highlight detail because I feel like it's pretty obscured and can affect users's ability to use certain functionality in the flyout or elsewhere.

Reviewer 2: It's difficult to answer, it would depend which flyout(or smth else) functionality we refer. 

As I mentioned  I think it is sufficient enough to point, source document will be merged to alert for non-aggregatin query. Depends on it, user would be able to use other Security Solution features.

I think scope of features too big to mention each single one

+ 
IMPORTANT: To turn on alert deduplication and allow query results to be associated with source events, you must add the `[metadata _id, _index, _version]` operator after the query's data source. Do not drop or filter out the `id`, `_index`, `_version` metadata properties in any part the query. 
+
Example non-aggregating query:  
+
[source,esql]
-----
FROM logs* [metadata _id, _index, _version]
| WHERE event.id == "8a4f500d"
| LIMIT 10
-----

> 

Reviewer 2: ```suggestion
WARNING:  Following examples drop _id field and are not valid for non-aggregating query

In this example command KEEP omits _id property, so it won't be returned in the result table
+
[source,esql]
-----
FROM logs* [metadata _id, _index, _version]
| WHERE event.category == "process"  AND event.id == "8a4f500d"
| KEEP event.*
| LIMIT 10

-----

In the next example command DROP explicitly omits  _id property, so it won't be returned in the result table
+
[source,esql]
-----
FROM logs* [metadata _id, _index, _version]
| WHERE event.category == "process"  AND event.id == "8a4f500d"
| DROP _id
| LIMIT 10

-----
```

Reviewer 2: added 2 examples to elaborate what dropping _id means

[[create-esql-rule]]
=== Create an ES|QL rule

preview::[]

Use {ref}/esql.html[ES|QL] to query your source events and aggregate event data. Query results are returned in a table with rows and columns. Each row becomes an alert.

To create an ES|QL rule:

> . Go to *Rules* -> *Detection rules (SIEM)* -> *Create new rule*. The *Create new rule* page displays.

Reviewer 3: ```suggestion
. Go to *Rules* -> *Detection rules (SIEM)* -> *Create new rule*. The *Create new rule* page appears.
```

> . Select **ES|QL**, then write a <<esql-rule-query-types,query>>. 

Reviewer 3: ```suggestion
. Select **ES|QL**, then write a <<esql-rule-query-types,query>>. 
```
Not sure the link is necessary here since it only jumps a few lines. Maybe it's just me but following it was a little disorienting

Author: Yeah, I've gone back and forth on this too. Other than implying that there's more to learn, it's not adding much value 😅

+
TIP: Click the help icon (image:images/esql-ref-button.png[Click the ES|QL help icon,20,20]) to open the in-product reference documentation for all ES|QL commands and functions.
+
. Click *Continue* to <<rule-ui-basic-params, configure basic rule settings>>.

[float]
[[esql-rule-query-types]]
==== ES|QL query types 

ES|QL rule queries are loosely categorized into two types: aggregating and non-aggregating. 

[float]
[[esql-agg-query]]
===== Aggregating query 

> An aggregating query uses {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate source event data. Generated alerts only contain the fields returned by the query.

Reviewer 3: ```suggestion
Aggregating queries use {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate source event data. The alerts they generate only contain the fields returned by the query.
```
If this is accurate I think it's a bit more clear, but I could be wrong!

Author: From a technical perspective, it might be more accurate to say that the rule generates alerts -- at least that's what I'm gathering from @vitaliidm's explanation [here](https://github.com/elastic/security-docs/pull/4062#discussion_r1374795992). 

How does this look, @benironside:
```suggestion
Aggregating queries use {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate source event data. Alerts generated by the {esql} rule only contain the fields returned by the query.
```

Reviewer 3: ```suggestion
Aggregating queries use {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate source event data. Alerts generated by an {esql} rule only contain the fields returned by its query.
```

Reviewer 3: What about this?

Here is an example aggregating query:

[source,esql]
----
FROM logs*
| STATS host_count = COUNT(host.name) BY host.name
| SORT host_count DESC
| WHERE host_count > 20
----

- This query starts by searching logs from indices that match the pattern `logs-*`. 
- The query then aggregates the count of events by `host.name`.
- Next, it sorts the result by `host_count` in descending order.

> - Then, it filters events where the `host_count` field is encountered more than 20 times during the specified rule interval.

Reviewer 3: ```suggestion
- Then, it filters for events where the `host_count` field appears more than 20 times during the specified rule interval.
```

> NOTE: Rules that use aggregating queries might create duplicate alerts. This can happen  when events that occur in the additional look-back time are aggregated twice: once in the current rule execution and once in the a previous rule execution.

Reviewer 3: ```suggestion
NOTE: Rules that use aggregating queries might create duplicate alerts. This can happen  when events that occur in the additional look-back time are aggregated both in the current rule execution and in a previous rule execution.
```

[float]
[[esql-non-agg-query]]
===== Non-aggregating query 

> A non-aggregating query doesn't use `STATS...BY` functions and doesn't aggregate source event data. Generated alerts contain source event data and fields returned by the ES|QL query. 

Reviewer 3: ```suggestion
Non-aggregating queries don't use `STATS...BY` functions and don't aggregate source event data. The alerts they generate contain source event data and fields returned by the ES|QL query. 
```

Author: Same thing that I mentioned [here](https://github.com/elastic/security-docs/pull/4062#discussion_r1378067372):

Non-aggregating queries don't use `STATS...BY` functions and don't aggregate source event data. Alerts generated by the {esql} rule only contain the fields returned by the query.

Author: Resolved in [c5e95f5](https://github.com/elastic/security-docs/pull/4062/commits/c5e95f5165a7816ca90f7ac80a3870e14fb89955).

> Here is an example of a non-aggregating query:

Reviewer 3: ```suggestion
Here is an example non-aggregating query:
```
Changed to match aggregating query example

[source,esql]
-----

> FROM logs* [metadata _id, _index, _version]

Reviewer 3: ```suggestion
FROM logs* [metadata _id, _index, _version]
```
Should this and the previous example say `logs-*` instead of `logs*` ? The description below says `logs-*` and that's what I would expect

Reviewer 3: This applies to all the examples below as well

Author: Good catch! Fixed in [cbc6444](https://github.com/elastic/security-docs/pull/4062/commits/cbc64440611a121f93d40cb3ad7d1b4f4c33dc10).

| WHERE event.category == "process"  AND event.id == "8a4f500d"
| LIMIT 10
-----
- This query starts by querying logs from indices that match the pattern `logs-*`. The `[metadata _id, _index, _version]` operator allows <<esql-non-agg-query-dedupe,alert deduplication>>.  
- Next, the query filters events where the `event.category` is a process and the `event.id` is `8a4f500d`. 

> - Then, it limits the output to the top 10 results.

Reviewer 3: top 10 results organized by what?

Author: Quick update: I'll address this once have more info on [this](https://github.com/elastic/security-docs/pull/4065#discussion_r1377546493).

Author: Resolving this as discussed in Slack.

[float]
[[esql-non-agg-query-dedupe]]

> ===== Deduplicating alerts for rules that use non-aggregating queries

Author: ```suggestion
===== Deduplicating alerts created by rules using non-aggregating queries
```

> To deduplicate alerts, the query needs access to the `_id`, `_index`, and `_version` metadata fields within the queried source event documents. You can allow this by adding the `[metadata _id, _index, _version]` operator after the `FROM` source command, for example:

Reviewer 3: ```suggestion
To deduplicate alerts, a query needs access to the `_id`, `_index`, and `_version` metadata fields of the queried source event documents. You can allow this by adding the `[metadata _id, _index, _version]` operator after the `FROM` source command, for example:
```

[source,esql]
-----
FROM logs* [metadata _id, _index, _version]
| WHERE event.category == "process"  AND event.id == "8a4f500d"
| LIMIT 10
-----

> Once the metadata fields are provided, a unique alert ID is created for each alert generated by the query.

Reviewer 3: ```suggestion
When those metadata fields are provided, unique alert IDs are created for each alert generated by the query.
```

> As you continue developing the query, make sure you don't {ref}/esql-commands.html#esql-drop[`DROP`] or filter out the `_id`, `_index`, or `_version` metadata fields. 

Reviewer 3: ```suggestion
When developing the query, make sure you don't {ref}/esql-commands.html#esql-drop[`DROP`] or filter out the `_id`, `_index`, or `_version` metadata fields. 
```

> Here is an example of an invalid query that uses the `DROP` command to omit the `_id` property from the results table:

Reviewer 3: Are the examples invalid, or would they just fail to deduplicate alerts?

Author: My understanding is that they'd fail to dedupe alerts. Is this version more clear?

```suggestion
Here is an example of a query that fails to deduplicate alerts. It uses the `DROP` command to omit the `_id` property from the results table:
```

Reviewer 3: Yes 100%. Looks great!

Reviewer 3: Yes 100%. Looks great!

To create an ES|QL rule:

. Go to *Rules* -> *Detection rules (SIEM)* -> *Create new rule*. The *Create new rule* page displays.
. Select **ES|QL**, then write a <<esql-rule-query-types,query>>. 
+
TIP: Click the help icon (image:images/esql-ref-button.png[Click the ES|QL help icon,20,20]) to open the in-product reference documentation for all ES|QL commands and functions.
+
. Click *Continue* to <<rule-ui-basic-params, configure basic rule settings>>.

[float]
[[esql-rule-query-types]]
==== ES|QL query types 

ES|QL rule queries are loosely categorized into two types: <<esql-agg-query,aggregating>> and <<esql-non-agg-query,non-aggregating>>. 

[float]
[[esql-agg-query]]
===== Aggregating query 

An aggregating query uses {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate source event data. Generated alerts only contain the fields returned by the query.

Here is an example aggregating query:

[source,esql]
----
FROM logs*
| STATS host_count = COUNT(host.name) BY host.name
| SORT host_count DESC
| WHERE host_count > 20
----

> This query generates an alert whenever the `host.name` field is encountered more than 20 times during the specified rule interval.

Author: @vitaliidm it might be helpful to break this example down a bit more so user can see where the aggregation happens. Would you mind verifying that the following information is correct:

```suggestion

- This query starts by querying logs from indices that match the pattern `logs-*`. 
- The query then aggregates the count of events by `host.name`.
- Next, it sorts the result by `host_count` in descending order.
- Then, it filters events where the `host_count` field is encountered more than 20 times during the specified rule interval.
```

Reviewer 2: looks good. thansl

[float]
[[esql-non-agg-query]]
===== Non-aggregating query 
A non-aggregating query doesn't use `STATS...BY` functions and doesn't aggregate source data. Generated alerts contain source event data and fields returned by the ES|QL query. 

> 

Author: Might be useful to add an example here as well, to keep with the structure of the previous section. Would you mind verifying this as well, @vitaliidm:

```suggestion
Here is an example non-aggregating query:

[source,esql]
-----
FROM logs* [metadata _id, _index, _version]
| WHERE event.category == "process"  AND event.id == "8a4f500d"
| KEEP event.*
| LIMIT 10
-----

- This query starts by querying logs from indices that match the pattern `logs-*`. The [metadata _id, _index, _version] operator is added to turn on alert deduplication.  
- Next, the query filters events where the `event.category` is a process and the `event.id` is `8a4f500d`. 
- It keeps the selected fields `event.*`.
- Then, it limits the output to the top 10 results.
```

Reviewer 2: I would phrase it as to allow deduplication, instead of "turn on".

Btw, this query is no valid, as it drops `_id` field. Should not it be a valid one?

Author: Ah yes, good catch. I'll remove that line from the query. How's this:

```suggestion
Here is an example non-aggregating query:
[source,esql]
-----
FROM logs* [metadata _id, _index, _version]
| WHERE event.category == "process"  AND event.id == "8a4f500d"
| LIMIT 10
-----
- This query starts by querying logs from indices that match the pattern `logs-*`. The `[metadata _id, _index, _version]` operator allows alert deduplication.  
- Next, the query filters events where the `event.category` is a process and the `event.id` is `8a4f500d`. 
- Then, it limits the output to the top 10 results.
```

> If you create a non-aggregating query, you also have to manually turn on alert deduplication and link source events to alerts. To do this, you must add the `[metadata _id, _index, _version]` operator after the `FROM` source command. For example:  

Author: Will be using this version, which we reviewed in Slack.

```suggestion
To deduplicates alerts, the query needs access to the _id, _index, and _version metadata fields within the queried source event documents.  You can allow this by adding the [metadata _id, _index, _version] operator after the FROM source command, for example:

[source,esql]
-----
FROM logs* [metadata _id, _index, _version]
| WHERE event.category == "process"  AND event.id == "8a4f500d"
| LIMIT 10
Once the metadata fields are provided, a unique alert ID is created for each alert generated by the query.

```

> If you create a non-aggregating query, you also have to manually turn on alert deduplication and link source events to alerts. To do this, you must add the `[metadata _id, _index, _version]` operator after the `FROM` source command. For example:  

Author: Both of these were added in [6ff4a1b](https://github.com/elastic/security-docs/pull/4062/commits/6ff4a1b530582ea52d354361a545455feb0d9388).

[source,esql]
-----
FROM logs* [metadata _id, _index, _version]
| WHERE event.id == "8a4f500d"

> | LIMIT 10

Reviewer 2: Just realised, LIMIT over rule max_signals won't have any effect on number of generated alerts. It will always pick smaller between:
`LIMIT` in ES|QL query and rule [max_signals](https://www.elastic.co/guide/en/security/current/rules-api-create.html#opt-fields-all) in rule's configuration

Author: This is a nice catch, and a good "gotcha" to note in the rule docs. This information applies to ES|QL rule queries in general, so it might make sense to place it under step 2 [here](https://github.com/elastic/security-docs/pull/4062/files#diff-bb6609d5341af025e47b175f8432a1ce9dd90d3c93d656dbc388c599b120924dR273). In the next iteration of these docs, maybe we can move it to a new section/page that provides additional query guidance and examples. 

It might also be a good idea to mention this information about the `LIMIT` and `max_signals` params in the `max_signals` [API docs](https://www.elastic.co/guide/en/security/current/rules-api-create.html#opt-fields-all). Just so there's coverage in both places and users see if whether they create an ES|QL rule from the UI or API. What do you think, @vitaliidm ?

Reviewer 2: Let's move it to step 2.

> It might also be a good idea to mention this information about the LIMIT and max_signals params in the max_signals [API docs](https://www.elastic.co/guide/en/security/current/rules-api-create.html#opt-fields-all). Just so there's coverage in both places and users see if whether they create an ES|QL rule from the UI or API. What do you think, @vitaliidm ?

I don't think we need to add it there. It's only related to ES|QL query.

Author: Quick update: requested language from Vitalii. Am keeping this thread open in the meantime.

Author: Chatted with Vitalii over Slack, and will need to add the following information:
- LIMIT defines how many rows will be returned, and because of this, the max number alerts per rule execution (`max_signals`)cannot be greater than what's defined in LIMIT.
- If LIMIT isn't present, `max_signals` will be taken in account.

Author: Added in [2d45097](https://github.com/elastic/security-docs/pull/4062/commits/2d450970274daa8dedf92385092f1d24f825615e).

[float]
[[esql-rule-query-types]]
==== ES|QL query types 

ES|QL rule queries are loosely categorized into two types: <<esql-agg-query,aggregating>> and <<esql-non-agg-query,non-aggregating>>. 

[float]
[[esql-agg-query]]
===== Aggregating query 

An aggregating query uses {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate source event data. Generated alerts only contain the fields returned by the query.

Here is an example aggregating query:

[source,esql]
----
FROM logs*
| STATS host_count = COUNT(host.name) BY host.name
| SORT host_count DESC
| WHERE host_count > 20
----

This query generates an alert whenever the `host.name` field is encountered more than 20 times during the specified rule interval.

NOTE: Aggregating queries might create duplicate alerts if events that occurred in a rule's additional look-back time are aggregated during the rule's current and previous rule execution cycles.

[float]
[[esql-non-agg-query]]
===== Non-aggregating query 
A non-aggregating query doesn't use `STATS...BY` functions and doesn't aggregate source data. Generated alerts contain source event data and fields returned by the ES|QL query. 

> To deduplicates alerts, the query needs access to the `_id`, `_index, and `_version` metadata fields within the queried source event documents.  You can allow this by adding the `[metadata _id, _index, _version]` operator after the FROM source command, for example:

Author: ```suggestion
To deduplicates alerts, the query needs access to the `_id`, `_index`, and `_version` metadata fields within the queried source event documents. You can allow this by adding the `[metadata _id, _index, _version]` operator after the FROM source command, for example:
```

+
. Click *Continue* to <<rule-ui-basic-params, configure basic rule settings>>.

[float]
[[esql-rule-query-types]]
==== ES|QL query types 

ES|QL rule queries are loosely categorized into two types: <<esql-agg-query,aggregating>> and <<esql-non-agg-query,non-aggregating>>. 

[float]
[[esql-agg-query]]
===== Aggregating query 

An aggregating query uses {ref}/esql-functions-operators.html#esql-agg-functions[`STATS...BY`] functions to aggregate source event data. Generated alerts only contain the fields returned by the query.

Here is an example aggregating query:

[source,esql]
----
FROM logs*
| STATS host_count = COUNT(host.name) BY host.name
| SORT host_count DESC
| WHERE host_count > 20
----

- This query starts by searching logs from indices that match the pattern `logs-*`. 
- The query then aggregates the count of events by `host.name`.
- Next, it sorts the result by `host_count` in descending order.
- Then, it filters events where the `host_count` field is encountered more than 20 times during the specified rule interval.

> NOTE: Aggregating queries might create duplicate alerts if events that occurred in a rule's additional look-back time are aggregated during the rule's current and previous rule execution cycles.

Reviewer 2: ```suggestion
NOTE: Agrregating query rules may generate duplicate alerts, when events in additional look-back time getting aggregated twice: in current and previous rule executions
```

queries do not create alerts, alerts getting created during rule execution. ES|QL query just return results, from which alerts are generated

Author: Oh yeah, I can see how this could be interpreted as ES|QL queries generating alerts. I'm not sure that we want to use the term "agrregating query rules" though. It could be read as a new/different rule type. How about this:

```suggestion
NOTE: Rules that use aggregating queries might create duplicate alerts. This can happen  when events that occur in the additional look-back time are aggregated twice: once in the current rule execution and once in the a previous rule execution.
```

Author: @vitaliidm I'm merging this for now, but we can revisit it in a couple weeks if needs to adjusted for technical accuracy.

Reviewer 2: agree, on aggr rquery rule terminlogy

[float]
[[esql-non-agg-query]]
===== Non-aggregating query 
A non-aggregating query doesn't use `STATS...BY` functions and doesn't aggregate source data. Generated alerts contain source event data and fields returned by the ES|QL query. 

> Here is an example non-aggregating query:

Reviewer 2: ```suggestion
Here is an example of non-aggregating query:
```

Author: We can go with the longer version of this if that's more clear. 

```suggestion
Here is an example of a non-aggregating query:
```
